design.md

```diff
@@ -0,0 +1,90 @@
+# Design Considerations:
+**Task:** port azuremetadata.pl (perl) to python
+
+**Requirements:**
+(Or rather, the Methodology used to Generate requirements):
+1. compare azuremetadata CLI with the other metadata CLI modules for other cloud providers (ec2 and gce)
+2. Identify and interview current users of metadata modules (Sean)
+
+**Observations:**
+- The 3 CLI metadata modules appear to have been written at 3 different times by 3 different
+authors.
+- Subtle and not so subtle differences exist in the CLI behaviors and output formats.
+This requires developers to implement separate handlers and be aware of the idiosyncrasies
+of each implementation.
+- The GCE and EC2 modules both use getopt() and are largely familiar API except for
+the returned results.
+- Azure (the target of this port) is expectedly weird with different API, results, and configuration parameters
+required (...and because it was written in Perl)
+- It looks like the list of available and supported CLI query options for each cloud provider
+has been added to over time.
+To allow a new item to be queried from the CLI, the cloud module must also
+be revised in order to access it.
+- There is also a practice of adding other SUSE specific functions into the "cloud metadata"
+rather than breaking them into separate CLIs. e.g "cloud-service", "billing-tag", etc.
+- The meta-data collection of CLIs seems primarily to support automation pipelines elsewhere
+in Enceladus and not meant exclusively for humans.
+As such these CLIs are expected to have consistent behavior across all the meta-data CLIs
+approaching that of an API rather than an ad-hoc query tool. Current best practice for API
```

Comment:
Well there are "inconsistencies" across clouds, this is unavoidable as framework providers choose their own naming, "location" vs. "region" for example. These will have to remain as these names are documented in the framework provider metadata documentation and we cannot and should not expect a user (command line or developer) to make the mental jumps to correlate names as they may appear in the tool vs. the way they are documented.